salt/states/x509_v2.py

```diff
@@ -0,0 +1,1685 @@
+
+
+This example state will instruct all minions to trust certificates signed by
+our new CA. Mind that this example works for Debian-based OS only.
+Also note the Jinja call to encode the string to JSON, which will avoid
+YAML issues with newline characters.
+
+.. code-block:: jinja
+
+    # /srv/salt/cert.sls
+
+    Ensure the CA trust bundle exists:
+      file.directory:
+        - name: /usr/local/share/ca-certificates
+
+    Ensure our self-signed CA certificate is included:
+      x509.pem_managed:
+        - name: /usr/local/share/ca-certificates/myca.crt
+        - text: {{ salt["mine.get"]("ca", "x509.get_pem_entries")["ca"]["/etc/pki/ca.crt"] | json }}
+
+This state creates a private key, then requests a certificate signed by our CA
+according to the www policy.
+
+.. code-block:: yaml
+
+    # /srv/salt/www.sls
+
+    Ensure PKI directory exists:
+      file.directory:
+        - name: /etc/pki
+
+    Create private key for the certificate:
+      x509.private_key_managed:
+        - name: /etc/pki/www.key
+        - keysize: 4096
+        - backup: true
+        - require:
+          - file: /etc/pki
+
+    Request certificate:
+      x509.certificate_managed:
+        - name: /etc/pki/www.crt
+        - ca_server: ca
+        - signing_policy: www
+        - private_key: /etc/pki/www.key
+        - CN: www.example.com
+        - days_remaining: 7
+        - backup: true
+        - require:
+          - x509: /etc/pki/www.key
+"""
+import base64
+import copy
+import datetime
+import logging
+import os.path
+
+import salt.utils.files
+from salt.exceptions import CommandExecutionError, SaltInvocationError
+from salt.state import STATE_INTERNAL_KEYWORDS as _STATE_INTERNAL_KEYWORDS
+
+try:
+    import cryptography.x509 as cx509
+    from cryptography.exceptions import UnsupportedAlgorithm
+    from cryptography.hazmat.primitives import hashes
+
+    import salt.utils.x509 as x509util
+
+    HAS_CRYPTOGRAPHY = True
+except ImportError:
+    HAS_CRYPTOGRAPHY = False
+
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = "x509"
+
+
+def __virtual__():
+    if not HAS_CRYPTOGRAPHY:
+        return (False, "Could not load cryptography")
+    if not __opts__.get("x509_v2"):
+        return (
+            False,
+            "x509_v2 needs to be explicitly enabled by setting `x509_v2: true` "
+            "in the minion configuration until Salt 3008 (Argon).",
+        )
+    return __virtualname__
+
+
+def certificate_managed(
+    name,
+    days_remaining=7,
+    ca_server=None,
+    signing_policy=None,
+    encoding="pem",
+    append_certs=None,
+    copypath=None,
+    prepend_cn=False,
+    digest="sha256",
+    signing_private_key=None,
+    signing_private_key_passphrase=None,
+    signing_cert=None,
+    public_key=None,
+    private_key=None,
+    private_key_passphrase=None,
+    csr=None,
+    subject=None,
+    serial_number=None,
+    not_before=None,
+    not_after=None,
+    days_valid=30,
```

Comment:
I think that's too short.

I would follow the example of LetsEncrypt for automated certificate management and default `days_valid` to 90 and `days_remaining` to 30.

Reply:
I'm fine with either, just think a year is much too long. I would add that LetsEncrypt has to compromise between security and scalability a lot more than one would on one's own infrastructure.

Maybe the team can express a preference here.

Reply:
a sane default would be what most companies use internally. which would be a year. however. as this is meant for automated refresh 30 days would be fine. either way it is a setting and can be changed if the end user doesn't like it.

Reply:
I vote 90 as it really doesn't matter since I will be overwriting the value anyway. Is this the only blocker from getting this merged? Willing to pitch in if there are any other outstanding issues on getting this merged in because Rocky 9 requires the use of the onedir install to work.